### PR TITLE
port cuda triangular solver to hip

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR ports cuda triangular solver to hip.
Move the `SolveStruct` to the `solver/(hip|cuda)` namespace in  `common_trs_kernels` to avoid the redefinition and the conflict of execution.

HIP does not support triangular solver solves matrix, so HIP uses `csrsv2` not `csrsm`.

### CODE REVIEW

File: hip/base/hipsparse_bindings.hip.hpp [hunk 1]
```diff
@@ -303,6 +303,128 @@ GKO_BIND_HIPSPARSE_CONJ_TRANSPOSE64(ValueType, detail::not_implemented);
...
+    inline void csrsv2_buffer_size(                                          \
+        hipsparseHandle_t handle, hipsparseOperation_t trans,                \
+        const size_type m, size_type nnz, const hipsparseMatDescr_t descr,   \
+        const ValueType *csrVal, const int32 *csrRowPtr,                     \
+        const int32 *csrColInd, csrsv2Info_t factor_info,                    \
+        int *factor_work_size)                                               \
+    {                                                                        \
+        GKO_ASSERT_NO_HIPSPARSE_ERRORS(HipsparseName(                        \
+            handle, trans, m, nnz, descr,                                    \
+            as_hiplibs_type(const_cast<ValueType *>(csrVal)), csrRowPtr,     \
```
Comment (upsj) [inline hip/base/hipsparse_bindings.hip.hpp]: Do we know why cuSPARSE/hipSPARSE want non-const access to the values? Why only in the `buffer_size` call?

File: cuda/solver/common_trs_kernels.cuh [hunk 2]
```diff
@@ -55,6 +55,94 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
...
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseSetMatDiagType(factor_descr, CUSPARSE_DIAG_TYPE_NON_UNIT));
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseCreateCsrsm2Info(&solve_info));
+        algorithm = 0;
+        policy = CUSPARSE_SOLVE_POLICY_USE_LEVEL;
+    }
+    SolveStruct(const SolveStruct &) : SolveStruct() {}
+    SolveStruct(SolveStruct &&) : SolveStruct() {}
+    SolveStruct &operator=(const SolveStruct &) { return *this; }
+    SolveStruct &operator=(SolveStruct &&) { return *this; }
```
Comment (upsj) [inline cuda/solver/common_trs_kernels.cuh]: It seems to me like we only pass around pointers to SolveStruct. Could we ` = delete` these move/copy-constructors/assignments instead? This behavior is very unintuitive.

File: hip/solver/common_trs_kernels.hip.hpp [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        policy = HIPSPARSE_SOLVE_POLICY_USE_LEVEL;
+    }
+
+    SolveStruct(const SolveStruct &) : SolveStruct() {}
+
+    SolveStruct(SolveStruct &&) : SolveStruct() {}
+
+    SolveStruct &operator=(const SolveStruct &) { return *this; }
+
+    SolveStruct &operator=(SolveStruct &&) { return *this; }
```
Comment (upsj) [inline hip/solver/common_trs_kernels.hip.hpp]: See cuda SolveStruct: Can we ` = delete` them?

File: hip/solver/common_trs_kernels.hip.hpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+
+template <typename ValueType, typename IndexType>
+void generate_kernel(std::shared_ptr<const HipExecutor> exec,
+                     const matrix::Csr<ValueType, IndexType> *matrix,
+                     solver::SolveStruct *solve_struct,
+                     const gko::size_type num_rhs, bool is_upper)
+{
+    auto hip_solve_struct =
+        reinterpret_cast<solver::hip::SolveStruct *>(solve_struct);
```
Comment (upsj) [inline hip/solver/common_trs_kernels.hip.hpp]: Shouldn't this be a `dynamic_cast`? Down-casting using `reinterpret_cast` could be very dangerous, as it ignores the data layout completely.

File: hip/solver/common_trs_kernels.hip.hpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                  const matrix::Csr<ValueType, IndexType> *matrix,
+                  const solver::SolveStruct *solve_struct,
+                  matrix::Dense<ValueType> *trans_b,
+                  matrix::Dense<ValueType> *trans_x,
+                  const matrix::Dense<ValueType> *b,
+                  matrix::Dense<ValueType> *x)
+{
+    using vec = matrix::Dense<ValueType>;
+    auto hip_solve_struct =
+        reinterpret_cast<const solver::hip::SolveStruct *>(solve_struct);
```
Comment (upsj) [inline hip/solver/common_trs_kernels.hip.hpp]: Same here

File: hip/solver/common_trs_kernels.hip.hpp [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    if (hipsparse::is_supported<ValueType, IndexType>::value) {
+        ValueType one = 1.0;
+        auto handle = exec->get_hipsparse_handle();
+
+        {
+            hipsparse::pointer_mode_guard pm_guard(handle);
+            if (b->get_stride() == 1) {
+                auto temp_b = const_cast<ValueType *>(b->get_const_values());
```
Comment (upsj) [inline hip/solver/common_trs_kernels.hip.hpp]: Why do we need this cast? Does hipSPARSE require non-const x?

File: cuda/solver/common_trs_kernels.cuh [hunk 7]
```diff
@@ -168,6 +261,8 @@ void solve_kernel(std::shared_ptr<const CudaExecutor> exec,
                   matrix::Dense<ValueType> *x)
 {
     using vec = matrix::Dense<ValueType>;
+    auto cuda_solve_struct =
+        reinterpret_cast<const solver::cuda::SolveStruct *>(solve_struct);
```
Comment (upsj) [inline cuda/solver/common_trs_kernels.cuh]: Same here: Potentially dangerous down-cast

File: hip/base/hipsparse_bindings.hip.hpp (hunk 1 above)
Comment (yhmtsai) [reply hip/base/hipsparse_bindings.hip.hpp]: I do not know why they use non-const value, but it should not change the value.

File: hip/solver/common_trs_kernels.hip.hpp (hunk 6 above)
Comment (yhmtsai) [reply hip/solver/common_trs_kernels.hip.hpp]: Good catch. No, we don't need that.
`csrsv2_solve`, `csrsm2_solve` use `const ValueType *`
`csrsm_solve` uses `ValueType *`

**Pull request merged**